src/ILLink.Tasks/CreateRuntimeRootDescriptorFile.cs

```diff
@@ -239,83 +295,130 @@ void OutputXml (string iLLinkTrimXmlFilePath, string outputFileName)
 			XmlDocument doc = new XmlDocument ();
 			doc.Load (iLLinkTrimXmlFilePath);
 			XmlNode linkerNode = doc["linker"];
+
+			if (featureSwitchMembers.Count > 0) {
+				foreach (var fsMembers in featureSwitchMembers.Keys) {
+					// <assembly fullname="System.Private.CoreLib" feature="System.Diagnostics.Tracing.EventSource.IsSupported" featurevalue="true" featuredefault="true">
+					XmlNode featureAssemblyNode = doc.CreateElement ("assembly");
+					XmlAttribute featureAssemblyFullName = doc.CreateAttribute ("fullname");
+					featureAssemblyFullName.Value = "System.Private.CoreLib";
+					featureAssemblyNode.Attributes.Append (featureAssemblyFullName);
+
+					XmlAttribute featureName = doc.CreateAttribute ("feature");
+					featureName.Value = fsMembers.Feature;
+					featureAssemblyNode.Attributes.Append (featureName);
+
+					XmlAttribute featureValue = doc.CreateAttribute ("featurevalue");
+					featureValue.Value = fsMembers.FeatureValue;
+					featureAssemblyNode.Attributes.Append (featureValue);
+
+					XmlAttribute featureDefault = doc.CreateAttribute ("featuredefault");
+					featureDefault.Value = fsMembers.FeatureDefault;
+					featureAssemblyNode.Attributes.Append (featureDefault);
+
+					foreach (var type in featureSwitchMembers[fsMembers]) {
+						AddXmlTypeNode (doc, featureAssemblyNode, type.Key, type.Value);
+					}
+					linkerNode.AppendChild (featureAssemblyNode);
+				}
+			}
+
 			XmlNode assemblyNode = linkerNode["assembly"];
 
 			foreach (string typeName in classNamesToClassMembers.Keys) {
-				XmlNode typeNode = doc.CreateElement ("type");
-				XmlAttribute typeFullName = doc.CreateAttribute ("fullname");
-				typeFullName.Value = typeName;
-				typeNode.Attributes.Append (typeFullName);
-
-				ClassMembers members = classNamesToClassMembers[typeName];
-
-				// We need to keep everyting in System.Runtime.InteropServices.WindowsRuntime and
-				// System.Threading.Volatile.
-				if (!typeName.StartsWith ("System.Runtime.InteropServices.WindowsRuntime") &&
-					!typeName.StartsWith ("System.Threading.Volatile")) {
-					if (members.keepAllFields) {
-						XmlAttribute preserve = doc.CreateAttribute ("preserve");
-						preserve.Value = "fields";
-						typeNode.Attributes.Append (preserve);
-					} else if ((members.fields == null) && (members.methods == null)) {
-						XmlAttribute preserve = doc.CreateAttribute ("preserve");
-						preserve.Value = "nothing";
-						typeNode.Attributes.Append (preserve);
-					}
+				AddXmlTypeNode (doc, assemblyNode, typeName, classNamesToClassMembers[typeName]);
```

Comment:
This is the second callsite - again the foreach above can be done over the dictionary - getting KeyValuePair and using that instead of repeating the lookup here.